analyzers/src/SonarAnalyzer.CSharp/Rules/ImplementISerializableCorrectly.cs

```diff
@@ -68,36 +68,46 @@ public class ImplementISerializableCorrectly : SonarDiagnosticAnalyzer
-                    implementationErrors.AddRange(CheckConstructor(typeKeyword, typeSymbol)); // Keyword
-                    implementationErrors.AddRange(CheckGetObjectDataAccessibility(typeSymbol, getObjectData));
-                    implementationErrors.AddRange(CheckGetObjectData(typeKeyword, typeSymbol, getObjectData)); // Keyword
+                    implementationErrors.AddRange(CheckSerializableAttribute(typeKeyword, typeSymbol));
+                    implementationErrors.AddRange(CheckConstructor(typeDeclarationSyntax, typeSymbol));
+                    implementationErrors.AddRange(CheckGetObjectDataAccessibility(typeDeclarationSyntax, typeSymbol, getObjectData));
+                    implementationErrors.AddRange(CheckGetObjectData(typeDeclarationSyntax, typeSymbol, getObjectData));
 
                     if (implementationErrors.Count > 0)
                     {
-                        c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, typeIdentifier.GetLocation(),
-                            additionalLocations: implementationErrors.ToAdditionalLocations(),
-                            properties: implementationErrors.ToProperties()));
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule,
+                                                                       typeIdentifier.GetLocation(),
+                                                                       additionalLocations: implementationErrors.ToAdditionalLocations(),
+                                                                       properties: implementationErrors.ToProperties()));
                     }
                 },
                 SyntaxKind.ClassDeclaration,
                 SyntaxKindEx.RecordDeclaration);
 
         private static IEnumerable<SecondaryLocation> CheckSerializableAttribute(SyntaxToken typeKeyword, INamedTypeSymbol typeSymbol)
         {
-            if (!HasSerializableAttribute(typeSymbol) &&
-                !typeSymbol.IsAbstract)
+            if (!HasSerializableAttribute(typeSymbol) && !typeSymbol.IsAbstract)
             {
                 yield return new SecondaryLocation(typeKeyword.GetLocation(),
-                    $"Add 'System.SerializableAttribute' attribute on '{typeSymbol.Name}' because it " +
-                    "implements 'ISerializable'.");
+                                                   $"Add 'System.SerializableAttribute' attribute on '{typeSymbol.Name}' because it implements 'ISerializable'.");
             }
         }
 
-        private static IEnumerable<TSyntax> GetDeclarations<TSyntax>(ISymbol symbol) =>
-            symbol == null ? Enumerable.Empty<TSyntax>() : symbol.DeclaringSyntaxReferences.Select(r => r.GetSyntax()).Cast<TSyntax>();
+        private static IEnumerable<TSyntax> DeclarationOrImplementation<TSyntax>(TypeDeclarationSyntax typeDeclaration, IMethodSymbol symbol)
+        {
+            if (symbol == null)
```

Comment:
this is called in three places:
- line 128 - `symbol` is null-checked in parent `if`
- line 154 - null-check is done at line 142
- line 181 - null-check done at line 174

So this cannot be `null` at this point, so you can remove it. It's a private method.

Eventually you can add a comment that it's expected to be null-checked in the caller.